WIP: Use systemd-cryptenroll with external TPM

ddstreet opened the pull request:
This is a rather large PR which primarily adds the ability to use systemd-cryptenroll to enroll a luks volume to an *external* TPM2 device.

This still needs two important things:
1) updated documentation (with examples)
2) tests to cover the added code

I wanted to open this PR now since this is a large patch series, and I'm sure there will be feedback requiring code changes, so i didn't want adding docs and tests to delay that feedback. However I don't think it should be merged until the docs and tests are added.


<!-- devel-freezer = {"comment-id":"1421407398","freezing-tag":"v253-rc2"} -->

ddstreet commented:
Well, the tests are using clang which has lambda functions instead of gcc's anonymous functions...I thought the coding style doc said we could assume C11 with gcc extensions :-(

I think I can rework it without needing anonymous functions.

berrange commented:
> This is a rather large PR which primarily adds the ability to use systemd-cryptenroll to enroll a luks volume to an _external_ TPM2 device.
> 
> This still needs two important things:
> 
>     1. updated documentation (with examples)

Until the docs are ready, could you just give a quick highlight of the usage of the new features as a comment here, as I think it'd help to understand what this series achieves when reviewing it.

ddstreet commented:
> > This is a rather large PR which primarily adds the ability to use systemd-cryptenroll to enroll a luks volume to an _external_ TPM2 device.
> > This still needs two important things:
> > ```
> > 1. updated documentation (with examples)
> > ```
> 
> Until the docs are ready, could you just give a quick highlight of the usage of the new features as a comment here, as I think it'd help to understand what this series achieves when reviewing it.

Sure; a quick example using 2 systems, "foo" and "bar":

1) on "bar", get a primary public key (in "tss" format, i.e. TPM2B_PUBLIC), and copy it to "foo":

```
root@bar:~# tpm2_createprimary -C o -c bar-primary.ctx
...
root@bar:~# tpm2_readpublic -c bar-primary.ctx -f tss -o bar-primary.tss
...
root@bar:~# scp bar-primary.tss foo:
...
```

2) on "foo", create a luks volume using a file, and enroll the tpm from "bar" (note the enrollment uses pcr policy only, with pcr 2, which must be identical on "foo" and "bar" since systemd-cryptenroll does not yet allow specifying expected pcr values):

```
root@foo:~# dd if=/dev/zero of=luks.img bs=1k count=64k
65536+0 records in
65536+0 records out
67108864 bytes (67 MB, 64 MiB) copied, 0.326964 s, 205 MB/s
root@foo:~# losetup -f --show luks.img 
/dev/loop0
root@foo:~# cryptsetup luksFormat /dev/loop0
...
root@foo:~# systemd-cryptenroll --tpm2-device auto --tpm2-pcrs 2 --tpm2-external-key bar-primary.tss /dev/loop0
🔐 Please enter current passphrase for disk /dev/loop0: ****                    
New TPM2 token enrolled as key slot 1.
```

3) copy the luks image to "bar":

```
root@foo:~# losetup -d /dev/loop0
root@foo:~# scp luks.img bar:
...
```

4) use systemd-cryptsetup to attach it:

```
root@bar:~# losetup -f --show luks.img 
/dev/loop0
root@bar:~# /lib/systemd/systemd-cryptsetup attach test /dev/loop0 - tpm2-device=auto
root@bar:~# 
```

ddstreet commented:
```
 ____________________________________________
/ Found   2 sanitizer errors (  2 ASan,   0  \
| UBSan,   0 MSan). Looks like you need to   |
\ look at the log                            /
 --------------------------------------------
 \
  \
     __
    /  \
    |  |
    @  @
    |  |
    || |/
    || ||
    |\_/|
    \___/

```

lol - thanks clippy!

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1367,9 +1367,9 @@ int tpm2_seal(const char *device,
         _cleanup_(Esys_Freep) TPM2B_PRIVATE *private = NULL;
         _cleanup_(Esys_Freep) TPM2B_PUBLIC *public = NULL;
         static const TPML_PCR_SELECTION creation_pcr = {};
+        _cleanup_(erase_and_freep) TPM2B_SENSITIVE_CREATE *hmac_sensitive = NULL;
```
see #26004 which solves this differently.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1410,7 +1410,7 @@ int tpm2_seal(const char *device,
                         .parameters.keyedHashDetail.scheme.scheme = TPM2_ALG_NULL,
-                        .unique.keyedHash.size = 32,
+                        .unique.keyedHash.size = SHA256_DIGEST_SIZE,
                         .authPolicy = *policy_digest,
```
will cherry-pick this into git main, no need to delay things for this

ddstreet replied:
thanks

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -695,15 +673,85 @@ static int tpm2_get_policy_digest(
+                size_t *ret_pcr_values_size) {
+
+        _cleanup_free_ TPM2B_DIGEST *pcr_values = NULL;
+        TPML_PCR_SELECTION remaining, read = {};
```
hmm, "read()" is a libc symbol. maybe pick a different name. old gcc used to complain about this, new doesn't anymore, but I#d still avoid it

ddstreet replied:
ack, changed it to 'total_read'

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -67,18 +88,31 @@ static inline void Esys_Freep(void *p) {
 
+#else /* HAVE_TPM2 */
+
+struct tpm2_context {};
```
hmm, why define it to an empty struct? it should suffic to define this one without declaring any actual definition, no, as before?

ddstreet replied:
ack, changed

poettering commented:
hmm, this needs a rebase already, and will need an even bigger one, once #25006 lands (which should be any minute now)

poettering commented:
so, hmm, first of all, could you rebase? 

but more importantly: can we split this up into multiple PRs? i.e. the handle handling rework could probably be a PR of its own which we can merge first

ddstreet commented on src/basic/fd-util.h:
```diff
@@ -90,14 +90,8 @@ static inline int make_null_stdio(void) {
-                _ret_;                          \
-        })
+/* Like TAKE_PTR() but for file descriptors, resetting them to -EBADF */
+#define TAKE_FD(fd) TAKE_GENERIC(fd, -EBADF)
```
updated in the latest push

ddstreet commented:
> so, hmm, first of all, could you rebase?

rebased, and addressed feedback

> 
> but more importantly: can we split this up into multiple PRs? i.e. the handle handling rework could probably be a PR of its own which we can merge first

no problem, though it will be monday or tues before i can get to it, but for sure will have it split into multiple PRs next week. thanks.

ddstreet commented:
Opened 2 PR's referenced above to start with; I'll work on splitting more out after review on those

ddstreet commented:
rebased on top of #26180 just to keep this fresh

williamcroberts commented:
What's an "external" tpm? What's the TL;DR of this patch?

bluca commented:
> What's an "external" tpm? What's the TL;DR of this patch?

This is about allowing to pre-seal against the TPM of another machine - eg, pre-encrypt and seal a confidential VM. So the goal is to be able to do the sealing without access to the TPM where the VM will run (and which will do the unsealing).

williamcroberts commented:
> > What's an "external" tpm? What's the TL;DR of this patch?
> 
> This is about allowing to pre-seal against the TPM of another machine - eg, pre-encrypt and seal a confidential VM. So the goal is to be able to do the sealing without access to the TPM where the VM will run (and which will do the unsealing).

So you perform the object protections on the seal data according to the storage parent and import into a TPM later?

bluca commented:
> > > What's an "external" tpm? What's the TL;DR of this patch?
> > 
> > 
> > This is about allowing to pre-seal against the TPM of another machine - eg, pre-encrypt and seal a confidential VM. So the goal is to be able to do the sealing without access to the TPM where the VM will run (and which will do the unsealing).
> 
> So you perform the object protections on the seal data according to the storage parent and import into a TPM later?

I think the strategy here is to get the destination TPM's public key and use that, but I'll let Dan specify the exact details as I'm only aware of the bigger picture

ddstreet commented:
> > > What's an "external" tpm? What's the TL;DR of this patch?
> > 
> > 
> > This is about allowing to pre-seal against the TPM of another machine - eg, pre-encrypt and seal a confidential VM. So the goal is to be able to do the sealing without access to the TPM where the VM will run (and which will do the unsealing).
> 
> So you perform the object protections on the seal data according to the storage parent and import into a TPM later?

The use case for this is in clouds, e.g. have a single 'provisioning' server that:

1) maintains a pool of pre-encrypted VM drives
2) on demand, generate a virtual TPM for a new instance
3) enroll one of the pre-encrypted VM drives to the newly generated TPM
4) pass the new instance (and its vTPM) off to a host system to boot it

The provisioning server doesn't ever actually boot the instance and doesn't have access to the running vTPM, only the SRK public key for the new instance.

williamcroberts commented:
> > > > What's an "external" tpm? What's the TL;DR of this patch?
> > > 
> > > 
> > > This is about allowing to pre-seal against the TPM of another machine - eg, pre-encrypt and seal a confidential VM. So the goal is to be able to do the sealing without access to the TPM where the VM will run (and which will do the unsealing).
> > 
> > 
> > So you perform the object protections on the seal data according to the storage parent and import into a TPM later?
> 
> I think the strategy here is to get the destination TPM's public key and use that, but I'll let Dan specify the exact details as I'm only aware of the bigger picture

Yeah that makes sense, grab the storage parent, add the object protections and then deploy to another TPM.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -307,308 +411,753 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                        &public);
+        if (rc != TSS2_RC_SUCCESS)
+                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                       "Failed to generate %s %skey in TPM: %s",
```
why wouldn't it be supported? i dont see anything in the spec about being optional

williamcroberts replied:
That command wasn't added until spec 1.38, and you'd be surprised at how many TPMs are on 1.16. You'll get bug reports, trust me: https://github.com/tpm2-software/tpm2-pkcs11/issues/264

williamcroberts replied:
It's best to call GetCap and see if it's supported and use it if it is and have a fallback for older TPM 2.0 firmware revisions.

ddstreet replied:
that's quite annoying :(

but thanks for the info; i'll update it

ddstreet commented:
rebasing to keep current

ddstreet commented:
Closing this, as it's mostly been split out into other PRs.
